File: Objective-C/Project-Model/Project-ObjectiveC/Project-ObjectiveC/SCLAlertViewPlus.m
//
//  SCLAlertViewPlus.m
//  AHK-100anos
//
//  Created by Erico GT on 10/7/16.
//  Copyright © 2016 Atlantic Solutions. All rights reserved.
//

#import "SCLAlertViewPlus.h"

@implementation SCLAlertViewPlus

- (instancetype)initWithNewWindow
{
    self = [super initWithNewWindow];
    return self;
}

- (instancetype)init
{
    self = [super initWithNewWindow];
    return self;
}

- (SCLButton *)addButton:(NSString *)title withType:(enumSCLAlertButtonType)type actionBlock:(SCLActionBlock)action
{
    SCLButton *button = [super addButton:title actionBlock:action];
    
    if (type == SCLAlertButtonType_Neutral){
        button.defaultBackgroundColor = [UIColor grayColor];
    }else if(type == SCLAlertButtonType_Question){
        button.defaultBackgroundColor = [UIColor colorWithRed:1.0/255.0 green:55.0/255.0 blue:91.0/255.0 alpha:1.0];
    }else if(type == SCLAlertButtonType_Destructive){
        button.defaultBackgroundColor = UIColorFromHEX(0xC1272D);
    }
    
    return button;
}

- (SCLButton *)addButton:(NSString *)title withType:(enumSCLAlertButtonType)type  validationBlock:(SCLValidationBlock)validationBlock actionBlock:(SCLActionBlock)action
{
    SCLButton *button = [super addButton:title validationBlock:validationBlock actionBlock:action];
    
    if (type == SCLAlertButtonType_Neutral){
        button.defaultBackgroundColor = [UIColor grayColor];
    }else if(type == SCLAlertButtonType_Question){
        button.defaultBackgroundColor = [UIColor colorWithRed:1.0/255.0 green:55.0/255.0 blue:91.0/255.0 alpha:1.0];
    }else if(type == SCLAlertButtonType_Destructive){
        button.defaultBackgroundColor = UIColorFromHEX(0xC1272D);
    }
    
    return button;
}

- (SCLButton *)addButton:(NSString *)title withType:(enumSCLAlertButtonType)type  target:(id)target selector:(SEL)selector
{
    SCLButton *button = [super addButton:title target:target selector:selector];
    
    if (type == SCLAlertButtonType_Neutral){
        button.defaultBackgroundColor = [UIColor grayColor];
    }else if(type == SCLAlertButtonType_Question){
        button.defaultBackgroundColor = [UIColor colorWithRed:1.0/255.0 green:55.0/255.0 blue:91.0/255.0 alpha:1.0];
    }else if(type == SCLAlertButtonType_Destructive){
        button.defaultBackgroundColor = UIColorFromHEX(0xC1272D);
    }
    
    return button;
}

+ (SCLAlertViewPlus*)createDefaultAlert
{
    SCLAlertViewPlus *alert = [[SCLAlertViewPlus alloc] initWithNewWindow];
    [alert.labelTitle setFont:[UIFont fontWithName:FONT_MYRIAD_PRO_SEMIBOLD size:20.0]];
    [alert.viewText setFont:[UIFont fontWithName:FONT_MYRIAD_PRO_REGULAR size:16.0]];
    [alert setButtonsTextFontFamily:FONT_MYRIAD_PRO_SEMIBOLD withSize:16.0];
    alert.showAnimationType = SCLAlertViewShowAnimationSlideInFromTop;//SlideInFromTop;
    alert.hideAnimationType = SCLAlertViewHideAnimationSlideOutToBottom;//SlideOutToBottom;
    alert.backgroundType = SCLAlertViewBackgroundShadow;
    return alert;
}

+ (SCLAlertViewPlus*)createRichAlertWithMessage:(NSString*)bodyMessage imagesArray:(NSArray<UIImage*>*)images animationTimePerFrame:(NSTimeInterval)aTime
{
    SCLAlertViewPlus *alert = [[SCLAlertViewPlus alloc] initWithNewWindow];
    [alert.labelTitle setFont:[UIFont fontWithName:FONT_MYRIAD_PRO_SEMIBOLD size:20.0]];
    [alert.viewText setFont:[UIFont fontWithName:FONT_MYRIAD_PRO_REGULAR size:16.0]];
    [alert setButtonsTextFontFamily:FONT_MYRIAD_PRO_SEMIBOLD withSize:16.0];
    alert.showAnimationType = SCLAlertViewShowAnimationSlideInFromTop;//SlideInFromTop;
    alert.hideAnimationType = SCLAlertViewHideAnimationSlideOutToBottom;//SlideOutToBottom;
    alert.backgroundType = SCLAlertViewBackgroundShadow;
    //
    CGFloat alertWidth = 216.0;
    CGRect textRect = [bodyMessage boundingRectWithSize:CGSizeMake(alertWidth, CGFLOAT_MAX)
                                         options:NSStringDrawingUsesLineFragmentOrigin| NSStringDrawingUsesFontLeading
                                      attributes:@{NSFontAttributeName:[UIFont fontWithName:FONT_MYRIAD_PRO_REGULAR size:16.0]}
                                         context:nil];
    
   
    //Label
    UILabel *label = [[UILabel alloc] initWithFrame:CGRectMake(0.0, 10.0, alertWidth, textRect.size.height)];
    label.textColor = [UIColor grayColor];
    label.font = [UIFont fontWithName:FONT_MYRIAD_PRO_REGULAR size:16.0];
    label.numberOfLines = 0;
    label.textAlignment = NSTextAlignmentCenter;
    label.text = bodyMessage;
    
    //ImageView
    UIImageView *imvView = [[UIImageView alloc] initWithFrame:CGRectMake(0.0, label.frame.size.height + 20.0, alertWidth, 120)];
    imvView.backgroundColor = [UIColor redColor];
    imvView.contentMode = UIViewContentModeScaleAspectFit;
    imvView.clipsToBounds = YES;
    imvView.animationImages = images;
    imvView.animationDuration = (aTime * (float)images.count);
    [imvView startAnimating];
    
    //CustomView
    UIView *subView = [[UIView alloc] initWithFrame:CGRectMake(0.0, 0.0, alertWidth, (10.0 + label.frame.size.height + 10.0 + imvView.frame.size.height + 20.0))];
    [subView addSubview:label];
    [subView addSubview:imvView];
    //
    [alert addCustomView:subView];
    
    return alert;
}

@end

File: Swift 4/Project-Model/Project-Swift/Project-Swift/SCLAlertViewPlus.swift
//
//  SCLAlertViewPlus.swift
//  Project-Swift
//
//  Created by Erico GT on 05/04/17.
//  Copyright © 2017 Atlantic Solutions. All rights reserved.
//

import SCLAlertView

//MARK: - Enums and Extensions

enum SCLAlertButtonType: Int {
    case Default         = 0
    case Neutral         = 1
    case Success         = 2
    case Error           = 3
    case Notice          = 4
    case Warning         = 5
    case Info            = 6
    case Edit            = 7
    case Wait            = 8
}

//MARK: - SCLAlertViewPlus
class SCLAlertViewPlus:SCLAlertView{
    
    required init() {
        
        let appearance = SCLAlertView.SCLAppearance(
            kTitleFont: UIFont.systemFont(ofSize: 20, weight: UIFont.Weight.medium),
            kTextFont: UIFont.systemFont(ofSize: 14, weight: UIFont.Weight.regular),
            kButtonFont: UIFont.systemFont(ofSize: 16, weight: UIFont.Weight.medium),
            showCloseButton: false
        )
        
        super.init(appearance: appearance)
    }
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
    }
    
    //MARK: - Class Funcs
    
    class func createDefaultAlert() -> SCLAlertViewPlus{
        
        let alert:SCLAlertViewPlus = SCLAlertViewPlus.init()
        //
        return alert
    }
    
    class func createRichAlert(bodyMessage:String, images:Array<UIImage>, animationTimePerFrame:TimeInterval) -> SCLAlertViewPlus{
        
        let alert:SCLAlertViewPlus = SCLAlertViewPlus.init()
        
        let font:UIFont = UIFont.systemFont(ofSize: 15, weight: UIFont.Weight.regular)
        
        let alertWidth:CGFloat = 216 //fixo pela superclass
        
        //TextField
        let label:UILabel = UILabel(frame: CGRect(x: 0.0, y: 10.0, width: alertWidth, height: bodyMessage.height(withConstrainedWidth: 216, font: font)))
        label.textColor = UIColor.gray
        label.font = font
        label.numberOfLines = 0
        label.textAlignment = NSTextAlignment.center
        label.text = bodyMessage
        
        //UIImageView
        let imageView = UIImageView(frame: CGRect(x: 0.0, y: (label.frame.size.height + 20.0), width: alertWidth, height: 100))
        imageView.backgroundColor = nil
        imageView.contentMode = UIViewContentMode.scaleAspectFit
        imageView.clipsToBounds = true
        //imageView.image = UIImage(named: "animal.jpg")
        imageView.animationImages = images
        imageView.animationDuration = animationTimePerFrame * Double(images.count)
        imageView.startAnimating()
        
        //CustomView
        let subview = UIView(frame: CGRect(x: 0.0, y: 0.0, width: alertWidth, height: (10.0 + label.frame.size.height + 10.0 + imageView.frame.size.height)))
        subview.addSubview(label)
        subview.addSubview(imageView)
        //
        alert.customSubview = subview
        
        return alert
    }
    
    //MARK: - Personalized Methods
    
    let appearance = SCLAlertView.SCLAppearance(
        showCircularIcon: false
    )
    
    func addButton(title:String, type:SCLAlertButtonType, action:@escaping () -> Void){
        
        var color:UIColor?
        
        switch type {
            case .Default:
                color = nil
                
            case .Neutral:
                color = UIColor.lightGray
            
            case .Success:
                color = self.UIColorFromRGB(0x22B573)
            
            case .Error:
                color = self.UIColorFromRGB(0xC1272D)
            
            case .Notice:
                color = self.UIColorFromRGB(0x727375)
            
            case .Warning:
                color = self.UIColorFromRGB(0xFFD110)
            
            case .Info:
                color = self.UIColorFromRGB(0x2866BF)
            
            case .Edit:
                color = self.UIColorFromRGB(0xA429FF)
            
            case .Wait:
                color = self.UIColorFromRGB(0xD62DA5)
        }

        super.addButton(title, backgroundColor: color, textColor: UIColor.white, showTimeout: nil) {
            action()
        }
    }
    
    //MARK: - Private Methods
    
    private func RGBFromUIColor(color:UIColor) -> UInt{
        
        var red: CGFloat = 0, green: CGFloat = 0, blue: CGFloat = 0, alpha: CGFloat = 0
        
        color.getRed(&red, green: &green, blue: &blue, alpha: &alpha)
        
        var colorAsUInt32 : UInt32 = 0
        colorAsUInt32 += UInt32(red * 255.0) << 16
        colorAsUInt32 += UInt32(green * 255.0) << 8
        colorAsUInt32 += UInt32(blue * 255.0)
        
        let colorAsUInt = UInt(colorAsUInt32)
        
        return colorAsUInt
    }
    
    private func UIColorFromRGB(_ rgbValue: UInt) -> UIColor {
        return UIColor(
            red: CGFloat((rgbValue & 0xFF0000) >> 16) / 255.0,
            green: CGFloat((rgbValue & 0x00FF00) >> 8) / 255.0,
            blue: CGFloat(rgbValue & 0x0000FF) / 255.0,
            alpha: CGFloat(1.0)
        )
    }

}
